### PR TITLE
[Gas Mechanics] Implement Gas history; Migrate to Arbitrum Gas model

### PR DESCRIPTION
**Pull request by StefanIliev545**
### Why this change is needed

This PR fixes the compatibility issues with metamask:
1) Transfering all the value of the account (or nearly all of it, its impossible to get it down to the exact amount due to l1 gas pricing; It's risky to modify the value of the message so this is the best it can get for now.)
2) The api that gets the account nonce was broken, which in turn made metamask get confused as to which transactions are successful and which aren't. 

Sim tests were changed:
There was a mix up in transfers tracking.
Gas pricing enabled.
Gas limit is now set by estimating the gas as the estimation now includes l1 fees.
Gas limit per batch raised significantly - The geth mempool rejects transactions with too high of a gas limit, but as the l1 cost is now hidden within the gas limit of a transaction, they can overrun the low ethereum cloned limits. Arbitrum has a very high gas limit and should (or at least was) use the same strategy we are.

### What changes were made as part of this PR

The L1 payments and refunds have been migrated on a per tx basis inside the EVM facade. 
Estimate gas takes a predicted l1 cost and divides it by the current l2 gas price in order to derive a gas quantity.

``Gas Quantity = Total Cost / Gas Price Per Unit``
 
Simulation tests have been adjusted.

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: tools/walletextension/container_run.sh [hunk 1]
```diff
@@ -11,9 +11,9 @@ set -euo pipefail
 port=3000
 portWS=3001
 host="0.0.0.0"
-nodeHost="erpc.sepolia-testnet.obscu.ro"
-nodePortHTTP=80
-nodePortWS=81
+nodeHost="host.docker.internal"
```
Comment (tudor-malene) [inline tools/walletextension/container_run.sh:14]: is this change intended?

File: tools/walletextension/README.md [hunk 2]
```diff
@@ -12,7 +12,7 @@ following commands from the `tools/walletextension/main` folder:
 
 ```
 env GOOS=darwin GOARCH=amd64 go build -o ../bin/wallet_extension_macos_amd64 .
-env GOOS=darwin GOARCH=arm64 go build -o ../bin/wallet_extension_macos_arm64 .
+    env GOOS=darwin GOARCH=arm64 go build -o ../bin/wallet_extension_macos_arm64 .
```
Comment (tudor-malene) [inline tools/walletextension/README.md:15]: just formatting

File: integration/simulation/transaction_injector.go [hunk 3]
```diff
@@ -141,10 +141,10 @@ func (ti *TransactionInjector) Start() {
 		})
 	}
 
-	wg.Go(func() error {
+	/*wg.Go(func() error {
```
Comment (tudor-malene) [inline integration/simulation/transaction_injector.go:144]: is this intended?

File: integration/networktest/tests/helpful/spin_up_local_network_test.go [hunk 4]
```diff
@@ -30,7 +30,6 @@ const (
 )
 
 func TestRunLocalNetwork(t *testing.T) {
-	networktest.TestOnlyRunsInIDE(t)
```
Comment (tudor-malene) [inline integration/networktest/tests/helpful/spin_up_local_network_test.go:33]: I think @BedrockSquirrel  had some reasoning to keep this not executing.

File: integration/common/constants.go [hunk 5]
```diff
@@ -83,7 +83,7 @@ func DefaultEnclaveConfig() *config.EnclaveConfig {
 		MaxBatchSize:              1024 * 25,
 		MaxRollupSize:             1024 * 64,
 		GasPaymentAddress:         gethcommon.HexToAddress("0xd6C9230053f45F873Cb66D8A02439380a37A4fbF"),
-		BaseFee:                   new(big.Int).SetUint64(1),
+		BaseFee:                   new(big.Int).SetUint64(1), //params.InitialBaseFee / 100),
```
Comment (tudor-malene) [inline integration/common/constants.go:86]: what's up with the comment

File: go/host/rpc/clientapi/client_api_eth.go [hunk 6]
```diff
@@ -184,15 +184,29 @@ func (api *EthereumAPI) GetStorageAt(_ context.Context, encryptedParams common.E
 
 // FeeHistory is a placeholder for an RPC method required by MetaMask/Remix.
 // rpc.DecimalOrHex -> []byte
-func (api *EthereumAPI) FeeHistory(context.Context, []byte, rpc.BlockNumber, []float64) (*FeeHistoryResult, error) {
+func (api *EthereumAPI) FeeHistory(context.Context, string, rpc.BlockNumber, []float64) (*FeeHistoryResult, error) {
 	// todo (#1621) - return a non-dummy fee history
+	header, err := api.host.DB().GetHeadBatchHeader()
+	if err != nil {
+		api.logger.Error(fmt.Sprintf("Unable to retrieve header for fee history. Cause: %w", err))
```
Comment (tudor-malene) [inline go/host/rpc/clientapi/client_api_eth.go:191]: no need for the sprintf

File: go/host/rpc/clientapi/client_api_eth.go [hunk 7]
```diff
@@ -184,15 +184,29 @@ func (api *EthereumAPI) GetStorageAt(_ context.Context, encryptedParams common.E
...
+	feeHist := &FeeHistoryResult{
+		OldestBlock:  (*hexutil.Big)(header.Number),
 		Reward:       [][]*hexutil.Big{},
 		BaseFee:      []*hexutil.Big{},
 		GasUsedRatio: []float64{},
-	}, nil
+	}
+
+	for _, header := range batches {
+		feeHist.GasUsedRatio = append(feeHist.GasUsedRatio, 0.9) //todo @siliev - fix when baseFee is dynamic
```
Comment (tudor-malene) [inline go/host/rpc/clientapi/client_api_eth.go:206]: what's this `0.9`?
Should it be a constant with a comment?

File: go/host/rpc/clientapi/client_api_eth.go [hunk 8]
```diff
@@ -184,15 +184,29 @@ func (api *EthereumAPI) GetStorageAt(_ context.Context, encryptedParams common.E
...
+		feeHist.BaseFee = append(feeHist.BaseFee, (*hexutil.Big)(header.BaseFee))
+	}
+	return feeHist, nil
```
Comment (tudor-malene) [inline go/host/rpc/clientapi/client_api_eth.go:209]: basically, you're returning the struct with the latest batch?

File: go/enclave/nodetype/sequencer.go [hunk 9]
```diff
@@ -234,6 +242,8 @@ func (s *sequencer) createNewHeadBatch(l1HeadBlock *common.L1Block, skipBatchIfE
 		}
 	}
 
+	sort.Sort(SortableTransactions(transactions))
```
Comment (tudor-malene) [inline go/enclave/nodetype/sequencer.go:245]: what's up with this?

**Review by tudor-malene — CHANGES_REQUESTED** [review]
Thanks Stefan. 
Quite a lot going on here.

Please add a description to the PR.

There are a few comments in the code, but I think it looks good overall

File: go/enclave/components/batch_executor.go [hunk 10]
```diff
@@ -93,18 +93,25 @@ func (executor *batchExecutor) payL1Fees(stateDB *state.StateDB, context *BatchE
...
 			continue
 		}
 		if accBalance.Cmp(cost) == -1 {
 			executor.logger.Info(fmt.Sprintf("insufficient account balance for tx - want: %d have: %d", cost, accBalance), log.TxKey, tx.Hash(), "addr", sender.Hex())
 			continue
 		}
-		stateDB.SubBalance(*sender, cost)
-		stateDB.AddBalance(context.Creator, cost)
+
+		/*stateDB.SubBalance(*sender, cost)
```
[the remark this answers is not in the document]
Comment (StefanIliev545) [reply go/enclave/components/batch_executor.go:107]: The fees are taken dynamically at the evm facade (and refunded) 
I use this only to forward the cost and filter out transactions, but this will probably be removed too, at some point later in time when its less destructive

File: go/enclave/enclave.go [hunk 11]
```diff
@@ -684,14 +686,28 @@ func (e *enclaveImpl) GetTransactionCount(encryptedParams common.EncryptedParams
...
+
 	// extract, create and validate the VK encryption handler
 	vkHandler, err := createVKHandler(&address, paramList[0], e.config.ObscuroChainID)
 	if err != nil {
 		return responses.AsPlaintextError(fmt.Errorf("unable to create VK encryptor - %w", err)), nil
 	}
 
 	var nonce uint64
-	l2Head, err := e.storage.FetchBatchBySeqNo(e.registry.HeadBatchSeq().Uint64())
+	l2Head, err := e.storage.FetchBatchBySeqNo(seqNo)
```
[the remark this answers is not in the document]
Comment (StefanIliev545) [reply go/enclave/enclave.go:710]: yep

File: go/enclave/evm/evm_facade.go [hunk 12]
```diff
@@ -67,51 +68,90 @@ func ExecuteTransactions(
...
-	receipt, err := gethcore.ApplyTransaction(cc, chain, nil, gp, s, header, t, usedGas, vmCfg)
 
-	// adjust the receipt to point to the right batch hash
-	if receipt != nil {
-		receipt.Logs = s.GetLogs(t.Hash(), batchHeight, batchHash)
-		receipt.BlockHash = batchHash
-		receipt.BlockNumber = big.NewInt(int64(batchHeight))
-		for _, l := range receipt.Logs {
-			l.BlockHash = batchHash
+	applyTx := func(
```
[the remark this answers is not in the document]
Comment (StefanIliev545) [reply go/enclave/evm/evm_facade.go:111]: It's a copy paste of gethcore.ApplyTransaction - can be moved externally.

File: go/host/rpc/clientapi/client_api_eth.go (hunk 8 above)
Comment (StefanIliev545) [reply go/host/rpc/clientapi/client_api_eth.go:209]: Yep

File: go/enclave/nodetype/sequencer.go (hunk 9 above)
Comment (StefanIliev545) [reply go/enclave/nodetype/sequencer.go:245]: Running the simulation it seemed like transactions werent ordered by nonces, but I'm not sure its necessary. The geth mempool is just a bit weird.

File: tools/walletextension/container_run.sh (hunk 1 above)
Comment (StefanIliev545) [reply tools/walletextension/container_run.sh:14]: Nope

File: integration/simulation/transaction_injector.go (hunk 3 above)
Comment (StefanIliev545) [reply integration/simulation/transaction_injector.go:144]: Will be uncommented again, just unclogging the sim tests step by step

File: go/enclave/components/batch_executor.go [hunk 13]
```diff
@@ -97,43 +97,25 @@ func (executor *batchExecutor) payL1Fees(stateDB *state.StateDB, context *BatchE
 		isFreeTransaction = isFreeTransaction && tx.GasPrice().Cmp(gethcommon.Big0) == 0
 
 		if isFreeTransaction {
-			freeTransactions = append(freeTransactions, tx)
+			freeTransactions = append(freeTransactions, common.L2PricedTransaction{
```
Comment (tudor-malene) [inline go/enclave/components/batch_executor.go:100]: I'm not quite clear what prevents a user from setting the "GasFeeCap=0", and become a "free transaction"?

File: go/enclave/components/batch_executor.go [hunk 14]
```diff
@@ -67,12 +67,12 @@ func NewBatchExecutor(
...
 }
 
-// payL1Fees - this function modifies the state db according to the transactions contained within the batch context
+// estimateL1Fees - this function modifies the state db according to the transactions contained within the batch context
 // in order to subtract gas fees from the balance. It returns a list of the transactions that have prepaid for their L1
 // publishing costs.
-func (executor *batchExecutor) payL1Fees(stateDB *state.StateDB, context *BatchExecutionContext) (common.L2Transactions, common.L2Transactions) {
-	transactions := make(common.L2Transactions, 0)
-	freeTransactions := make(common.L2Transactions, 0)
+func (executor *batchExecutor) estimateL1Fees(stateDB *state.StateDB, context *BatchExecutionContext) (common.L2PricedTransactions, common.L2PricedTransactions) {
```
Comment (tudor-malene) [inline go/enclave/components/batch_executor.go:73]: also looks to me like this method is excluding transactions that don't have enough money to pay for the l1 cost. 
I think the name and comment should reflect its responsibilities a bit better

File: go/enclave/components/batch_executor.go [hunk 15]
```diff
@@ -183,30 +165,36 @@ func (executor *batchExecutor) ComputeBatch(context *BatchExecutionContext, fail
...
+	xchainTxs := make(common.L2PricedTransactions, 0)
+	for _, xTx := range crossChainTransactions {
+		xchainTxs = append(xchainTxs, common.L2PricedTransaction{
+			Tx:             xTx,
+			PublishingCost: big.NewInt(0),
+		})
+	}
 
-	crossChainTransactions = append(crossChainTransactions, freeTransactions...)
+	xchainTxs = append(xchainTxs, freeTransactions...)
```
Comment (tudor-malene) [inline go/enclave/components/batch_executor.go:178]: shouldn't this be logically the other way round? "freeTransactions" is the array that contains xChainTxs and other stuff.
I think these variables need more clear names and comments

File: go/enclave/components/batch_executor.go [hunk 16]
```diff
@@ -230,6 +218,7 @@ func (executor *batchExecutor) ComputeBatch(context *BatchExecutionContext, fail
 
 	// the logs and receipts produced by the EVM have the wrong hash which must be adjusted
 	for _, receipt := range allReceipts {
+		executor.logger.Info("New receipt! ", log.TxKey, receipt.TxHash)
```
Comment (tudor-malene) [inline go/enclave/components/batch_executor.go:221]: this will be incredibly noisy

File: go/enclave/evm/evm_facade.go [hunk 17]
```diff
@@ -68,51 +72,90 @@ func ExecuteTransactions(
...
+		if err != nil {
+			return nil, err
+		}
+		l1cost := tx.PublishingCost
+		l1Gas := big.NewInt(0)
+		hasL1Cost := l1cost.Cmp(big.NewInt(0)) != 0
+
+		if hasL1Cost {
+			l1Gas.Div(l1cost, header.BaseFee)
+			l1Gas.Add(l1Gas, big.NewInt(1))
```
Comment (tudor-malene) [inline go/enclave/evm/evm_facade.go:136]: 1 - magic number. What does it represetn?

File: go/enclave/evm/evm_facade.go [hunk 18]
```diff
@@ -68,51 +72,90 @@ func ExecuteTransactions(
...
+	) (*types.Receipt, error) {
+		msg, err := gethcore.TransactionToMessage(tx.Tx, types.MakeSigner(config, header.Number, header.Time), header.BaseFee)
+		if err != nil {
+			return nil, err
+		}
+		l1cost := tx.PublishingCost
+		l1Gas := big.NewInt(0)
+		hasL1Cost := l1cost.Cmp(big.NewInt(0)) != 0
+
+		if hasL1Cost {
```
Comment (tudor-malene) [inline go/enclave/evm/evm_facade.go:134]: this block needs a fat comment

File: go/enclave/gas/oracle.go [hunk 19]
```diff
@@ -49,3 +52,15 @@ func (o *oracle) EstimateL1StorageGasCost(tx *types.Transaction, block *types.Bl
...
 	return big.NewInt(0).Mul(l1Gas, block.BaseFee()), nil
 }
+
+func (o *oracle) EstimateL1CostForMsg(args *gethapi.TransactionArgs, block *types.Block) (*big.Int, error) {
+	encoded := make([]byte, 0)
+	if args.Data != nil {
+		encoded = append(encoded, *args.Data...)
+	}
+
+	nonZeroGas := big.NewInt(int64(params.TxDataNonZeroGasEIP2028))
```
Comment (tudor-malene) [inline go/enclave/gas/oracle.go:62]: what do these things mean?

File: go/enclave/l2chain/l2_chain.go [hunk 20]
```diff
@@ -133,7 +133,7 @@ func (oc *obscuroChain) ObsCallAtBlock(apiArgs *gethapi.TransactionArgs, blockNu
 		return nil, fmt.Errorf("unable to fetch head state batch. Cause: %w", err)
 	}
 
-	callMsg, err := apiArgs.ToMessage(batch.Header.GasLimit, batch.Header.BaseFee)
+	callMsg, err := apiArgs.ToMessage(batch.Header.GasLimit-1, batch.Header.BaseFee)
```
Comment (tudor-malene) [inline go/enclave/l2chain/l2_chain.go:136]: why -1?

File: go/enclave/nodetype/sequencer.go (hunk 9 above)
Comment (tudor-malene) [reply go/enclave/nodetype/sequencer.go:247]: Umm. I'd say this deserves a bit of investigation. Maybe instead of sorting, check that what the mempool produces is sorted by nonce. And if not, we should investigate why not. It's probably some misconfiguration

File: go/host/rpc/clientapi/client_api_eth.go [hunk 21]
```diff
@@ -184,15 +185,29 @@ func (api *EthereumAPI) GetStorageAt(_ context.Context, encryptedParams common.E
...
+	feeHist := &FeeHistoryResult{
+		OldestBlock:  (*hexutil.Big)(header.Number),
 		Reward:       [][]*hexutil.Big{},
 		BaseFee:      []*hexutil.Big{},
 		GasUsedRatio: []float64{},
-	}, nil
+	}
+
+	for _, header := range batches {
+		feeHist.GasUsedRatio = append(feeHist.GasUsedRatio, 0.9) // todo @siliev - fix when baseFee is dynamic
```
Comment (tudor-malene) [inline go/host/rpc/clientapi/client_api_eth.go:207]: what is the significance of the 0.9 magic number?

**Review by tudor-malene — CHANGES_REQUESTED** [review]
Needs a few more comments.

**Review by tudor-malene — CHANGES_REQUESTED** [review]
needs a few more comments

File: go/enclave/components/batch_executor.go (hunk 13 above)
Comment (StefanIliev545) [reply go/enclave/components/batch_executor.go:100]: The mempool would reject such a transaction. The only way to get it in is bypass the mempool.

File: go/enclave/l2chain/l2_chain.go (hunk 20 above)
Comment (StefanIliev545) [reply go/enclave/l2chain/l2_chain.go:136]: *crickets* ... I must have had some genius thought that is now forever lost 🪕

File: go/enclave/components/batch_executor.go [hunk 22]
```diff
@@ -175,19 +175,19 @@ func (executor *batchExecutor) ComputeBatch(context *BatchExecutionContext, fail
 		})
 	}
 
-	xchainTxs = append(xchainTxs, freeTransactions...)
+	syntheticTransactions := append(xchainTxs, freeTransactions...)
```
Comment (tudor-malene) [inline go/enclave/components/batch_executor.go:178]: much clearer

**Review by tudor-malene — APPROVED** [review]
lgtm
